### DISCUSSION TITLE
TMC Connection error

### DISCUSSION THREAD

**Discussion by msdossajo** [Jyers/Marlin#317] [Q&A] [answered]
Hello,
I've just modified my Ender 3 V2 (4.2.2) motherboard to use Linear Advance and compiled a new firmware.
When I issue the command M122 I get information like "Set current", "StealthChop false" but also get the message "Testing E connection... Error: All LOW" and the screen shows the message "TMC CONNECTION ERROR".

Does anyone here is using this modification to use Linear Advance?
[M122.txt](https://github.com/Jyers/Marlin/files/6192045/M122.txt)

Thanks

**Comment by Reggi3** [comment, marked as answer]
Did you connect the 24v power to the board?  When I did this mod (a few days ago) I thought I'd be able to test it by powering the board via the USB serial connection and nothing else plugged in, as if it were a standard arduino board.  This was wrong, you need to have the 24v power connected to the board otherwise the driver chips don't respond, any data you are seeing from an M122 command is probably being populated via [Configuration_adv.h](https://github.com/Jyers/Marlin/blob/Ender_3_V2_Extensible_UI/Configuration%20Files/E3V2%20Templates/BLTouch-3x3-HighSpeed/Configuration_adv.h#L2445)


I haven't fully tested this mod yet, I still need some questions answered but I can confirm that Jyers firmware does compile for TMC Uart mods and the firmware, along with the tmc2208 drivers(<---Not a typo) respond correctly to M122, M569 and M906 commands.

**Reply by msdossajo** [reply]
@Reggi3 yes, I was using the 24v. The motherboard is back to it's place and is working, still get the message but I can use Linear Advance. Another interesting thing is that I had to double the number of steps, apparently the mod changed the micro-step setting in the driver.
After seen some information about the UART connection to the TMC2208 I'm not sure if using the pin PA3 for TX/RX is the best option to use the Trinamic debug. Probably one day I will hijack the PB13 (Beeper pin) to have TX and RX in different pins :-)

Before the mod I was one of the "lucky" owners that could use Linear Advance, now the only diference is that the stepper motor is much quieter using Linear Advance.

**Reply by Reggi3** [reply]
@msdossajo Interesting, I did read in the instructions for the Linear Advance mod that sometimes the connection to PA3 is a bit unstable.

It's odd that you mention having to double the steps, as they are all hard wired to 16 steps on the PCB and if you refer back to the link in my previous post, you can see that all tmc2208 uart drivers also default to 16 steps, I wonder if the issue is related to interpolation being turned on?

I double checked this myself because I have a 4.2.2 board that I've done the full mod on and I wanted to check that the 4.2.7 board behaves in the same way and it does, it's hard wired to High/Low = 1/0 on MS1/MS2 (1/1 would = 32 steps on the 4.2.7 board as it uses TMC2225 chips), whereas the 4.2.2 board is hardwired High/High = 1/1 on MS1/MS2 pins, so in theory, switching to TMC uart mode shouldn't impact a working system.

if you want to try some other pins, checkout the 4 pin header next to the LCD socket, the 2 pins in the middle are PA13 (SWDIO) and PA14 (SWCLK), both unused in the marlin firmware, both of them 100% work as TMC uart ports!!

```
Send: M122
Recv:     X  Y  Z  E
Recv: Enabled    false  false  false  false
Recv: Set current  850  780  850  990
Recv: RMS current  826  764  826  939
Recv: MAX current  1165  1077  1165  1324
Recv: Run current  26/31  24/31  26/31  16/31
Recv: Hold current  13/31  12/31  13/31  8/31
Recv: CS actual  13/31  12/31  13/31  8/31
Recv: PWM scale
Recv: vsense    1=.18  1=.18  1=.18  0=.325
Recv: stealthChop  false  false  false  false
Recv: msteps    16  16  16  16
Recv: interp    true  true  true  true
Recv: tstep    max  max  max  max
Recv: PWM thresh.
Recv: [mm/s]
Recv: OT prewarn  false  false  false  false
Recv: pwm scale sum  15  14  15  10
Recv: pwm scale auto  0  0  0  0
Recv: pwm offset auto  36  36  36  36
Recv: pwm grad auto  14  14  14  14
Recv: off time  0  0  0  0
Recv: blank time  24  24  24  24
Recv: hysteresis
Recv:  -end    2  2  2  2
Recv:  -start    1  1  1  1
Recv: Stallguard thrs
Recv: uStep count  8  8  8  40
Recv: DRVSTATUS  X  Y  Z  E
Recv: sg_result
Recv: stst
Recv: olb
Recv: ola
Recv: s2gb
Recv: s2ga
Recv: otpw
Recv: ot
Recv: 157C
Recv: 150C
Recv: 143C
Recv: 120C
Recv: s2vsa
Recv: s2vsb
Recv: Driver registers:
Recv:     X  0x80:0D:00:00
Recv:     Y  0x80:0C:00:00
Recv:     Z  0x80:0D:00:00
Recv:     E  0x80:08:00:00
Recv: 
Recv: 
Recv: Testing X connection... OK
Recv: Testing Y connection... OK
Recv: Testing Z connection... OK
Recv: Testing E connection... OK
Recv: ok
[...]


```

There are also 2 other physical pins that you might be able to use too, 1 is PA7 (Z endstop, it's unused if you use a BLTouch) and PA4 (Filament runout sensor).  I have tested PA7 and it works, I haven't tested PA4 yet though.

Lastly, if you don't have a BLTouch, then I *think* that will give 2 spare pins as well but I haven't tested that theory myself.

you might also want to change this line in [configuration_adv.h](https://github.com/Jyers/Marlin/blob/Ender_3_V2_Extensible_UI/Configuration%20Files/Jyers's%20Config/Configuration_adv.h#L2636)

Change from   
#define CHOPPER_TIMING CHOPPER_DEFAULT_12V        // All axes (override below)
to
#define CHOPPER_TIMING CHOPPER_DEFAULT_24V        // All axes (override below)

As our motors run at 24v.

**Comment by msdossajo** [comment]
@Reggi3 thank you for remember me that there are some free pins on the board, probably will go for the Z enstop.
Will use the following cable to connect, keeping the PA3 as RX and use PA7 to TX.
![FSVCUMPJHATJSN0](https://user-images.githubusercontent.com/72068472/112951743-97bc6d00-9133-11eb-8a42-85d8092b9836.png)
